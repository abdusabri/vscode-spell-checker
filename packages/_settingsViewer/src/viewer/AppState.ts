import {observable, computed} from 'mobx';
import { Settings, ConfigTarget, LocalId, SettingByConfigTarget, Config, Configs, LocalList, WorkspaceFolder, TextDocument, configTargets, ConfigTargets } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';


type Maybe<T> = T | undefined;
type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigTarget;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>,
    target: ConfigTarget
}

type InheritedFromTarget<T> = {
    value: Exclude<T, undefined>,
    target: ConfigTarget;
}


type InheritMembers<T> = {
    [K in keyof T]: InheritedFromTarget<T[K]>;
}

type InheritedConfigs = SettingByConfigTarget<Config>;

export class AppState implements State {
    @observable activeTabName = '';
    @observable settings: Settings = {
        dictionaries: [
        ],
        configs: {
            user: { inherited: {}, locals: [], languageIdsEnabled: [] },
            workspace: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            folder: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            file: undefined,
        }
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {}

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter(target => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        return tabs.filter(tab => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find(t => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locals = config.locals; // todo: calc inheritance
            const inherited = config.inherited.locals;

            const infos = new Map<string, LanguageInfo>();

            const addLocalsToInfos = (locals: string[], dictionaryName: string | undefined) => {
                locals.map(normalizeCode).map(lookupCode).filter(notUndefined).forEach(lang => {
                    const { code, lang: language, country } = lang;
                    const name = country ? `${language} - ${country}` : language;
                    const found = this.isLocalEnabledEx(target, code);
                    const enabled = found && found.value || false;
                    const info: LanguageInfo = infos.get(name) || {
                        code,
                        name,
                        enabled,
                        dictionaries: [],
                    };
                    if (dictionaryName) {
                        info.dictionaries.push(dictionaryName);
                    }
                    infos.set(name, info);
                });
            }
            if (locals) {
                addLocalsToInfos(locals, undefined);
            }
            this.settings.dictionaries.forEach(dict => addLocalsToInfos(dict.locals, dict.name));

            return {
                languages: [...infos.values()].sort(compareEach(
                    compareBy(info => !info.dictionaries.length),
                    compareBy('name'),
                )),
                inherited
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        }
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex(t => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = workspace && workspace.workspaceFolders || [];
        return folders.map(f => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.workspaceFolders || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.textDocuments || [];
    }

    @computed get documentSelection(): { label: string, value: string }[] {
        return this.documents.map(doc => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.name === name)[0];
    }

    actionSetLocal(target: ConfigTarget, local: LocalId, enable: boolean) {
        const localsCurrent = this.settings.configs[target].locals;
        const locals = (enable ? [...localsCurrent, local] : localsCurrent.filter(v => v !== local)).filter(uniqueFilter());
        this.settings.configs[target].locals = locals;
        this.messageBus.postMessage({ command: 'EnableLocalMessage', value: { target, local, enable } });
    }

    actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocalId):InheritedFromTarget<boolean> {
        const config = this.settings.configs[field];
        const target = config.inherited.locals || field;
        const locals = config.locals;
        return  {
            value: locals.map(normalizeCode).includes(code),
            target
        };
    }


    actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    actionActivateTab(tabName: string) {
        this.activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri })
        }
    }

    actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (fileConfig && fileConfig.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target }});
    }
}

function notUndefined<T>(a : T): a is Exclude<T, undefined> {
    return a !== undefined;
}


