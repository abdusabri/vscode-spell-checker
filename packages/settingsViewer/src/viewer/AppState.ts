import {observable, computed} from 'mobx';
import { Settings } from '../api/settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable settings: Settings = {
        locals: [
            {
                code: 'en',
                name: 'English',
                dictionaries: ['en', 'en-us'],
                enabled: true,
                isInUserSettings: true,
            },
            {
                code: 'es',
                name: 'Spanish',
                dictionaries: ['es', 'es-ES'],
                enabled: true,
                isInUserSettings: true,
            },
        ],
    };

    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
