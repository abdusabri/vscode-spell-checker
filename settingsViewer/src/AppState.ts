import {observable, computed} from 'mobx';
import { ConfigurationForDocument } from './settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable config: ConfigurationForDocument = {
        languageEnabled: true,
        fileEnabled: true,
        settings: {},
        docSettings: {}
    };
    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
