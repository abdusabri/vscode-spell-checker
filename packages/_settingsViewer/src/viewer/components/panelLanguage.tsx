import * as React from 'react';
import {Cell, Grid, Row} from '@material/react-layout-grid';
import {Checkbox} from '@material/react-checkbox';
import { AppState } from '../AppState';
import * as spellCheckIcon from '../images/SpellCheck.xs.png';
import {tf} from '../../api/utils';
import { LocalInfo } from '../AppState';
import { LocalSetting, LocalId } from '../../api/settings';

type LocalSettingField = keyof LocalSetting;


export function checkboxLocalInfo(appState: AppState, code: LocalId, field: LocalSettingField) {
    const isEnabled = appState.isLocalEnabled(field, code);
    const checked = isEnabled || false;
    const indeterminate =  isEnabled === undefined;
    return <Checkbox
        checked={checked}
        indeterminate={indeterminate}
        onChange={(e) => {
            const checked = e.target.indeterminate ? false : e.target.checked;
            appState.setLocal(field, code, checked);
        }}
        initRipple={() => {}}
    ></Checkbox>;
}

export class LanguagePanel extends React.Component<{appState: AppState}, {}> {
    render() {
        const appState = this.props.appState;
        const checkbox = (local: LocalInfo, field: LocalSettingField) =>
            checkboxLocalInfo(appState, local.code, field);
        const locals = appState.locals.map((local, index) => <Row key={index}>
            <Cell columns={3}>{local.name}<br/>{local.code}</Cell>
            <Cell columns={3}>{local.dictionaries.join(', ')}</Cell>
            <Cell columns={2}>
                {checkbox(local, 'user')}
            </Cell>
            <Cell columns={2}>
                {checkbox(local, 'workspace')}
            </Cell>
            <Cell columns={1}>
                {checkbox(local, 'folder')}
            </Cell>
            <Cell columns={1}>{tf(local && local.enabled)}</Cell>
        </Row>);
        return (
            <Grid>
                <Row>
                    <Cell columns={12}>
                        <img src={spellCheckIcon} />
                    </Cell>
                </Row>
                <Row>
                    <Cell columns={3}>Language</Cell>
                    <Cell columns={3}>dictionaries</Cell>
                    <Cell columns={2}>Global</Cell>
                    <Cell columns={2}>Workspace</Cell>
                    <Cell columns={1}>Folder</Cell>
                    <Cell columns={1}>Enabled</Cell>
                </Row>
                {locals}
            </Grid>
        );
     }
}
