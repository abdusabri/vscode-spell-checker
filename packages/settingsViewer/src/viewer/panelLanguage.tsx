import * as React from 'react';
import {Cell, Grid, Row} from '@material/react-layout-grid';
import {Checkbox} from '@material/react-checkbox';
import { AppState } from './AppState';
import * as spellCheckIcon from './images/SpellCheck.xs.png';
import {tf} from '../api/utils';
import { LocalInfo } from '../api/settings';

type OptionalBool = boolean | undefined;
type PropertyNamesOfTypeS<T, S> = { [K in keyof T]: T[K] extends S ? K : never }[keyof T];
type BooleanKeyOfLocalInfo = Exclude<PropertyNamesOfTypeS<LocalInfo, OptionalBool>, undefined>;


export function checkboxLocalInfo(appState: AppState, local: LocalInfo, index: number, field: BooleanKeyOfLocalInfo) {
    return <Checkbox
        checked={!!local[field]}
        indeterminate={local[field] === undefined}
        onChange={(e) => {
            const v = e.target.indeterminate ? undefined : e.target.checked;
            appState.settings.locals[index][field] = v;
        }}
        initRipple={() => {}}
    ></Checkbox>;
}

export class LanguagePanel extends React.Component<{appState: AppState}, {}> {
    render() {
        const appState = this.props.appState;
        const settings = appState.settings;
        const checkbox = (local: LocalInfo, index: number, field: BooleanKeyOfLocalInfo) =>
            checkboxLocalInfo(appState, local, index, field);
        const locals = settings.locals.map((local, index) => <Row key={index}>
            <Cell columns={3}>{local.name}</Cell>
            <Cell columns={3}>{local.dictionaries.join(', ')}</Cell>
            <Cell columns={2}>
                {checkbox(local, index, 'isInUserSettings')}
            </Cell>
            <Cell columns={2}>
                {checkbox(local, index, 'isInWorkspaceSettings')}
            </Cell>
            <Cell columns={1}>
                {checkbox(local, index, 'isInFolderSettings')}
            </Cell>
            <Cell columns={1}>{tf(local.enabled)}</Cell>
        </Row>);
        return (
            <Grid>
                <Row>
                    <Cell columns={12}>
                        <img src={spellCheckIcon} />
                    </Cell>
                </Row>
                <Row>
                    <Cell columns={3}>Language</Cell>
                    <Cell columns={3}>dictionaries</Cell>
                    <Cell columns={2}>Global</Cell>
                    <Cell columns={2}>Workspace</Cell>
                    <Cell columns={1}>Folder</Cell>
                    <Cell columns={1}>Enabled</Cell>
                </Row>
                {locals}
            </Grid>
        );
     }
}
