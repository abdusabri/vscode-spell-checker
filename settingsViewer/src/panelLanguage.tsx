import * as React from 'react';
import {Cell, Grid, Row} from '@material/react-layout-grid';
import {Checkbox} from '@material/react-checkbox';
import { AppState } from './AppState';
import * as spellCheckIcon from './images/SpellCheck.xs.png';
import {tf} from './utils';

export class LanguagePanel extends React.Component<{appState: AppState}, {}> {
    render() {
        const appState = this.props.appState;
        const settings = appState.settings;
        const locals = settings.locals.map((local, index) => <Row>
            <Cell columns={4}>{local.name}</Cell>
            <Cell columns={4}>{local.dictionaries.join(', ')}</Cell>
            <Cell columns={1}>
                <Checkbox
                    checked={!!local.isInUserSettings}
                    indeterminate={local.isInUserSettings === undefined}
                    onChange={(e) => {
                        const v = e.target.indeterminate ? undefined : e.target.checked;
                        appState.settings.locals[index].isInUserSettings = v;
                        appState.settings.locals[index].isInFolderSettings = v;
                    }}
                    initRipple={(e) => {}}
                ></Checkbox>
            </Cell>
            <Cell columns={1}>
                <Checkbox
                    checked={!!local.isInWorkspaceSettings}
                    indeterminate={local.isInWorkspaceSettings === undefined}
                    onChange={(e) => {
                        const v = e.target.indeterminate ? undefined : e.target.checked;
                        appState.settings.locals[index].isInWorkspaceSettings = v;
                    }}
                    initRipple={(e) => {}}
                ></Checkbox>
            </Cell>
            <Cell columns={1}>
                <Checkbox
                    checked={!!local.isInFolderSettings}
                    indeterminate={local.isInFolderSettings === undefined}
                    onChange={(e) => {
                        const v = e.target.indeterminate ? undefined : e.target.checked;
                        appState.settings.locals[index].isInFolderSettings = v;
                    }}
                    initRipple={(e) => {}}
                ></Checkbox>
            </Cell>
            <Cell columns={1}>{tf(local.enabled)}</Cell>
        </Row>);
        return (
            <Grid>
                <Row>
                    <Cell columns={12}>
                        <img src={spellCheckIcon} />
                    </Cell>
                </Row>
                {locals}
            </Grid>
        );
     }
}
