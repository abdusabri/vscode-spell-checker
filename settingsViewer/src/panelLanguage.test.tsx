
import * as React from 'react';
import { LanguagePanel, checkboxLocalInfo } from './panelLanguage';
import { AppState } from './AppState';
import { create } from 'react-test-renderer';

describe('Language Panel Verification', () => {
    it('checkboxLocalInfo', () => {
        const appState: AppState = new AppState();
        const checkbox = create(checkboxLocalInfo(appState, appState.settings.locals[0], 0, 'isInFolderSettings'));
        expect(checkbox.toJSON()).toMatchSnapshot();
        const instance = checkbox.getInstance()!;
        instance.props['onChange']({target: { checked: false, indeterminate: false }});
        expect(appState.settings.locals[0].isInFolderSettings).toBe(false);
        instance.props['onChange']({target: { checked: true, indeterminate: false }});
        expect(appState.settings.locals[0].isInFolderSettings).toBe(true);
        instance.props['onChange']({target: { checked: true, indeterminate: true }});
        expect(appState.settings.locals[0].isInFolderSettings).toBeUndefined();
    });

    it('tests the snapshot', () => {
        const appState: AppState = new AppState();
        const panelRenderer = create(<LanguagePanel appState={appState}></LanguagePanel>).toJSON()!;
        expect(panelRenderer).toMatchSnapshot();
    });
});
