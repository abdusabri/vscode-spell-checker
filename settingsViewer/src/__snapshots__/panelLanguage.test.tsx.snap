// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language Panel Verification checkboxLocalInfo 1`] = `
<div
  className="mdc-checkbox mdc-checkbox--upgraded"
  onAnimationEnd={[Function]}
>
  <input
    aria-checked="mixed"
    checked={false}
    className="mdc-checkbox__native-control"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    className="mdc-checkbox__background"
  >
    <svg
      className="mdc-checkbox__checkmark"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        className="mdc-checkbox__checkmark-path"
        d="M1.73,12.91 8.1,19.28 22.79,4.59"
        fill="none"
      />
    </svg>
    <div
      className="mdc-checkbox__mixedmark"
    />
  </div>
</div>
`;

exports[`Language Panel Verification tests the snapshot 1`] = `
<div
  className="mdc-layout-grid"
>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
    >
      <img
        src="test-file-stub"
      />
    </div>
  </div>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-4"
    >
      Language
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-2"
    >
      dictionaries
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-2"
    >
      Global
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-2"
    >
      Workspace
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-2"
    >
      Folder
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      Enabled
    </div>
  </div>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-4"
    >
      English
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-4"
    >
      en, en-us
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      <div
        className="mdc-checkbox mdc-checkbox--upgraded"
        onAnimationEnd={[Function]}
      >
        <input
          aria-checked={false}
          checked={true}
          className="mdc-checkbox__native-control"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      <div
        className="mdc-checkbox mdc-checkbox--upgraded"
        onAnimationEnd={[Function]}
      >
        <input
          aria-checked="mixed"
          checked={false}
          className="mdc-checkbox__native-control"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      <div
        className="mdc-checkbox mdc-checkbox--upgraded"
        onAnimationEnd={[Function]}
      >
        <input
          aria-checked="mixed"
          checked={false}
          className="mdc-checkbox__native-control"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      true
    </div>
  </div>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-4"
    >
      Spanish
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-4"
    >
      es, es-ES
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      <div
        className="mdc-checkbox mdc-checkbox--upgraded"
        onAnimationEnd={[Function]}
      >
        <input
          aria-checked={false}
          checked={true}
          className="mdc-checkbox__native-control"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      <div
        className="mdc-checkbox mdc-checkbox--upgraded"
        onAnimationEnd={[Function]}
      >
        <input
          aria-checked="mixed"
          checked={false}
          className="mdc-checkbox__native-control"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      <div
        className="mdc-checkbox mdc-checkbox--upgraded"
        onAnimationEnd={[Function]}
      >
        <input
          aria-checked="mixed"
          checked={false}
          className="mdc-checkbox__native-control"
          disabled={false}
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="mdc-checkbox__background"
        >
          <svg
            className="mdc-checkbox__checkmark"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              className="mdc-checkbox__checkmark-path"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
              fill="none"
            />
          </svg>
          <div
            className="mdc-checkbox__mixedmark"
          />
        </div>
      </div>
    </div>
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-1"
    >
      true
    </div>
  </div>
</div>
`;
